asynciterator.ts

```diff
@@ -1214,6 +1218,92 @@ function destinationFillBuffer<S>(this: InternalSource<S>) {
     (this._destination as any)._fillBuffer();
 }
 
+export class MappingIterator<S, D = S> extends AsyncIterator<D> {
+  constructor(source: AsyncIterator<S>, map: (item: S) => D) {
+    super();
+    let item: S | null;
+    this.read = (): D | null => {
+      if ((item = source.read()) !== null)
+        return map.call(this, item);
+      return null;
+    };
+    source.on('end', () => {
+      this.close();
+    });
+    source.on('readable', () => {
+      this.readable = true;
+    });
+  }
+}
+
+export class FilteringIterator<T> extends AsyncIterator<T> {
+  constructor(source: AsyncIterator<T>, filter: (item: T) => boolean) {
+    super();
+    let item: T | null;
+    this.read = (): T | null => {
+      while ((item = source.read()) !== null) {
+        if (filter.call(this, item))
+          return item;
+      }
+      return null;
+    };
+    source.on('end', () => {
+      this.close();
+    });
+    source.on('readable', () => {
+      this.readable = true;
+    });
+  }
+}
+
+export class SkippingIterator<T> extends AsyncIterator<T> {
+  constructor(source: AsyncIterator<T>, skip: number) {
+    super();
+    let item: T | null;
+    let skipped = 0;
+    this.read = (): T | null => {
+      while ((item = source.read()) !== null) {
```

Comment:
Rather than having all of this logic once skipping is finished why not modify the `read` method once you are done skipping?

```ts
   this.read = (): T | null => {
      while ((item = source.read()) !== null) {
        if (skipped < skip)
          skipped += 1;
        else {
             this.read = () => source.read();
             return item;
        }
      }
      return null;
    };
```

Reply:
Also doesn't holding the item in the global scope mean that you are unnecessarily persisting it in memory after it has been read?

Reply:
> why not modify the read method once you are done skipping

Agreed! I do something similar here https://github.com/belayeng/quadstore/blob/2a538b729df63351dc587bc15b83cc7283cba9d2/lib/get/sortingIterator.ts#L45 . However, setting methods in the constructor and/or messing with prototypal methods in the constructor is often frowned upon in TS codebases and I don't know where @RubenVerborgh stands on this.

> Also doesn't holding the item in the global scope mean that you are unnecessarily persisting it in memory after it has been read?

Yes. This is a habit of mine that dates back to when doing so actually made for a performance difference on some platforms. By now I think every platform must have optimized that away and this is probably just a bad habit to have.

Reply:
Monkey patching methods is indeed not a good idea in general.